### ISSUE TITLE
Consider current (.NET 4.0-era) guidance for unmanaged resources 

### ISSUE THREAD

**Issue by jnm2** [dotnet/runtime#43975]
Analyzer package: [Microsoft.CodeAnalysis.FxCopAnalyzers](https://www.nuget.org/packages/Microsoft.CodeAnalysis.FxCopAnalyzers)
Package Version: 2.6.2-beta2
Diagnostic ID: CA1063

Since .NET 4.0 was released in 2010, the guidance has been to use SafeHandle instead of writing finalizers and Dispose(bool) methods. The unmanagedness is fully encapsulated within the SafeHandle class, so the class that uses the SafeHandle no longer needs a finalizer or a Dispose(bool) method.

Unfortunately, the MSDN documentation has continued encouraging disposal of unmanaged resources to be put in Dispose(bool) methods, perpetuating a pattern which is verbose and has fewer safety guarantees.

For extremely rare cases where the unmanaged resource is not represented by an IntPtr handle, it's still safer to represent the unmanaged resource as a private class deriving from CriticalFinalizerObject than to store the fields directly inside the class that uses them.

(@sharwell, @davkean, may I use you as references since I've seen you speak out about this?)

Therefore, I believe this warning pushes folks in the wrong direction:

> CA1063 Provide an overridable implementation of Dispose(bool) on ServicesIntegrationTests or mark the type as sealed. A call to Dispose(false) should only clean up native resources. A call to Dispose(true) should clean up both managed and native resources.

Even if the class is virtual, we can assume that any derived class will also not need a finalizer or Dispose(bool) method because best practice is for the derived class to encapsulate the unmanagedness in a SafeHandle class rather than to store unmanaged references as direct fields.

**Comment by sharwell** [comment]
The last time this came up, the conclusion is CA1063 still contains the official framework design guidance for public APIs. As with the other framework design guidelines, this rule is primarily intended for improving the quality of shared APIs. While it is often convenient for projects to adopt the same rules for non-public code (consistency/simplicity/maintainability), it is allowed for internal code to deviate from the guidelines while still adhering to their _primary_ intent.

Other `IDisposable` patterns are not part of the official guidance, so it is _at best_ a non-goal of the framework design analyzers to support them. However, if a proposal came to somehow tweak the rules of the analyzers and/or split the rules such that they were more convenient for authors, it might be possible to accommodate the request without undermining the primary functionality.

With that said, my _personal_ view certainly deviates from the official design guidelines. I've found the simplest solution to be disabling CA1063 and other `IDisposable` rules in my own libraries.

**Comment by NN---** [comment]
Agree, I don't have any finalizer in the code and don't suggest anyone to deal with it manually.
Everything is wrapped in SafeHandle or other managed wrapper, therefore I don't have any direct use of unmanaged resource.

All of my Dispose methods are simply calling members Dispose and this rule is useless in such case.

**Comment by buyaa-n** [comment]
@mavasani @sharwell looks the [dispose guideline is updated](https://github.com/dotnet/docs/pull/18266). 

We want to determine what needs to be done for the issue.  Seems you are suggesting to disable the CA1063 analyzer for .NetCore and propose a new analyzer for using SafeHandle instead of implementing a finalizer or a Dispose(bool) methods for a class. What should we do next?
